File: src/lib/capacity.ts
import type { Prisma } from "@prisma/client";
import { BLOCKING_STATUSES, DAY_BUCKET } from "./constants";
import { db } from "./db";
import { isUniqueConstraint } from "./booking-errors";
import { memberMatchesServiceKind, parseRoles } from "./roles";

export type CapacityMember = {
  id: string;
  name: string;
  status: string;
  roles?: string;
  dailyCapacity: number;
  services: Array<{ serviceId: string }>;
};

export type CapacityBusiness = {
  members: CapacityMember[];
  shifts: Array<{ id: string; active: boolean; capacity: number | null }>;
  services?: Array<{ id: string; kind: string }>;
};

export function activeMembers(business: CapacityBusiness) {
  return business.members.filter((member) => member.status === "ACTIVE");
}

export function memberCanPerform(member: CapacityMember, serviceId: string, kind?: string) {
  if (kind) return memberMatchesServiceKind(parseRoles(member.roles), kind);
  return member.services.some((row) => row.serviceId === serviceId);
}

type Tx = Prisma.TransactionClient | typeof db;

export class CapacityFullError extends Error {
  constructor(message = "CAPACITY_FULL") {
    super(message);
    this.name = "CapacityFullError";
  }
}

export class PreferredUnavailableError extends Error {
  constructor() {
    super("PREFERRED_UNAVAILABLE");
    this.name = "PreferredUnavailableError";
  }
}

export type StaffSnapshot = {
  id: string;
  name: string;
  dailyCapacity: number;
  shiftCapacity: number | null;
  serviceIds: string[];
  remainingDay: number;
  remainingShift: number;
};

export function remainingOf(staff: StaffSnapshot) {
  return Math.max(0, Math.min(staff.remainingDay, staff.remainingShift));
}

export async function usedSeats(
  tx: Tx,
  args: { teamMemberId: string; date: string; bucket: string },
) {
  return tx.capacityHold.count({
    where: { teamMemberId: args.teamMemberId, date: args.date, bucket: args.bucket },
  });
}

export async function staffSnapshots(
  tx: Tx,
  business: CapacityBusiness,
  date: string,
  shiftId: string | null,
): Promise<StaffSnapshot[]> {
  const shift = shiftId ? business.shifts.find((row) => row.id === shiftId && row.active) : null;
  const members = activeMembers(business);
  const rows: StaffSnapshot[] = [];
  for (const member of members) {
    const usedDay = await usedSeats(tx, { teamMemberId: member.id, date, bucket: DAY_BUCKET });
    const usedShift = shift
      ? await usedSeats(tx, { teamMemberId: member.id, date, bucket: shift.id })
      : 0;
    const shiftCap = shift?.capacity ?? member.dailyCapacity;
    rows.push({
      id: member.id,
      name: member.name,
      dailyCapacity: member.dailyCapacity,
      shiftCapacity: shift?.capacity ?? null,
      serviceIds: member.services.map((row) => row.serviceId),
      remainingDay: Math.max(0, member.dailyCapacity - usedDay),
      remainingShift: Math.max(0, shiftCap - usedShift),
    });
  }
  return rows;
}

export async function claimCapacitySeats(
  tx: Tx,
  args: {
    date: string;
    shiftId: string | null;
    bookingId: string;
    memberIds: string[];
    members: Array<{ id: string; dailyCapacity: number }>;
    shiftCapacity: number | null;
  },
) {
  const uniqueIds = [...new Set(args.memberIds)];
  for (const memberId of uniqueIds) {
    const member = args.members.find((row) => row.id === memberId);
    if (!member) throw new CapacityFullError();
    await claimBucket(tx, {
      teamMemberId: memberId,
      date: args.date,
      bucket: DAY_BUCKET,
      capacity: member.dailyCapacity,
      bookingId: args.bookingId,
    });
    if (args.shiftId) {
      await claimBucket(tx, {
        teamMemberId: memberId,
        date: args.date,
        bucket: args.shiftId,
        capacity: args.shiftCapacity ?? member.dailyCapacity,
        bookingId: args.bookingId,
      });
    }
  }
}

async function claimBucket(
  tx: Tx,
  args: { teamMemberId: string; date: string; bucket: string; capacity: number; bookingId: string },
) {
  if (args.capacity <= 0) throw new CapacityFullError();
  const existing = await tx.capacityHold.findMany({
    where: { teamMemberId: args.teamMemberId, date: args.date, bucket: args.bucket },
    select: { seat: true },
  });
  const taken = new Set(existing.map((row) => row.seat));
  for (let seat = 0; seat < args.capacity; seat += 1) {
    if (taken.has(seat)) continue;
    try {
      await tx.capacityHold.create({
        data: {
          teamMemberId: args.teamMemberId,
          date: args.date,
          bucket: args.bucket,
          seat,
          bookingId: args.bookingId,
        },
      });
      return;
    } catch (error) {
      if (isUniqueConstraint(error)) continue;
      throw error;
    }
  }
  throw new CapacityFullError();
}

export async function releaseCapacityHolds(tx: Tx, bookingId: string) {
  await tx.capacityHold.deleteMany({ where: { bookingId } });
}

export type ServiceAssignment = { serviceId: string; teamMemberId: string };

function sanitizePreferredByService(raw: Record<string, string> | null | undefined, serviceIds: string[]) {
  const map: Record<string, string> = {};
  if (!raw) return map;
  for (const serviceId of serviceIds) {
    const memberId = raw[serviceId]?.trim();
    if (memberId) map[serviceId] = memberId;
  }
  return map;
}

export function assignStaff(args: {
  serviceIds: string[];
  staff: StaffSnapshot[];
  preferredMemberId?: string | null;
  preferredByService?: Record<string, string> | null;
  business: CapacityBusiness;
}): ServiceAssignment[] {
  const remaining = new Map(args.staff.map((row) => [row.id, remainingOf(row)]));
  const assignedThisBooking = new Set<string>();
  const result: ServiceAssignment[] = [];

  const kindOf = (serviceId: string) => args.business.services?.find((row) => row.id === serviceId)?.kind;

  const capable = (serviceId: string) =>
    args.staff.filter((row) => {
      const member = args.business.members.find((item) => item.id === row.id);
      return member && memberCanPerform(member, serviceId, kindOf(serviceId));
    });

  function takeMember(memberId: string, serviceId: string) {
    if (!assignedThisBooking.has(memberId)) {
      const left = remaining.get(memberId) ?? 0;
      if (left < 1) throw new CapacityFullError();
      remaining.set(memberId, left - 1);
      assignedThisBooking.add(memberId);
    }
    result.push({ serviceId, teamMemberId: memberId });
  }

  const preferredMap = sanitizePreferredByService(args.preferredByService, args.serviceIds);
  const preferred = args.preferredMemberId
    ? args.staff.find((row) => row.id === args.preferredMemberId)
    : null;
  if (args.preferredMemberId && !preferred) throw new PreferredUnavailableError();
  if (preferred) {
    for (const serviceId of args.serviceIds) {
      const member = args.business.members.find((item) => item.id === preferred.id);
      if (!preferredMap[serviceId] && member && memberCanPerform(member, serviceId, kindOf(serviceId))) {
        preferredMap[serviceId] = preferred.id;
      }
    }
  }

  const leftover: string[] = [];
  for (const serviceId of args.serviceIds) {
    const preferredId = preferredMap[serviceId];
    if (!preferredId) {
      leftover.push(serviceId);
      continue;
    }
    const snapshot = args.staff.find((row) => row.id === preferredId);
    const bizMember = args.business.members.find((item) => item.id === preferredId);
    if (!snapshot || !bizMember || !memberCanPerform(bizMember, serviceId, kindOf(serviceId))) {
      throw new PreferredUnavailableError();
    }
    if (remainingOf(snapshot) < 1 && !assignedThisBooking.has(snapshot.id)) {
      throw new PreferredUnavailableError();
    }
    try {
      takeMember(snapshot.id, serviceId);
    } catch (error) {
      if (error instanceof CapacityFullError) throw new PreferredUnavailableError();
      throw error;
    }
  }

  for (const serviceId of leftover) {
    const options = capable(serviceId);
    if (!options.length) throw new CapacityFullError("NO_STAFF");
    const ranked = [...options].sort((a, b) => {
      const ra = assignedThisBooking.has(a.id) ? 999 : remaining.get(a.id) ?? 0;
      const rb = assignedThisBooking.has(b.id) ? 999 : remaining.get(b.id) ?? 0;
      return rb - ra || a.name.localeCompare(b.name);
    });
    const pick = ranked.find((row) => assignedThisBooking.has(row.id) || (remaining.get(row.id) ?? 0) > 0);
    if (!pick) throw new CapacityFullError();
    takeMember(pick.id, serviceId);
  }

  return result;
}

export function remainingBookingsForServices(
  business: CapacityBusiness,
  staff: StaffSnapshot[],
  serviceIds: string[],
) {
  let count = 0;
  const clone: StaffSnapshot[] = staff.map((row) => ({ ...row }));
  for (let i = 0; i < 40; i += 1) {
    try {
      const assigned = assignStaff({ serviceIds, staff: clone, business });
      const used = new Set(assigned.map((row) => row.teamMemberId));
      for (const id of used) {
        const row = clone.find((item) => item.id === id);
        if (!row) continue;
        row.remainingDay = Math.max(0, row.remainingDay - 1);
        row.remainingShift = Math.max(0, row.remainingShift - 1);
      }
      count += 1;
    } catch {
      break;
    }
  }
  return count;
}

export function blockingWhere(date: string): Prisma.BookingWhereInput {
  return { date, status: { in: [...BLOCKING_STATUSES] } };
}

File: backend/bridey_api/capacity.py
from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.orm import Session

from bridey_api.constants import DAY_BUCKET
from bridey_api.errors import CapacityFullError, PreferredUnavailableError, is_unique_constraint
from bridey_api.ids import new_id
from bridey_api.models import CapacityHold


def claim_capacity_seats(
    session: Session,
    *,
    date: str,
    shift_id: str | None,
    booking_id: str,
    member_ids: list[str],
    members: list[dict],
    shift_capacity: int | None,
) -> None:
    unique_ids = list(dict.fromkeys(member_ids))
    for member_id in unique_ids:
        member = next((row for row in members if row["id"] == member_id), None)
        if not member:
            raise CapacityFullError()
        _claim_bucket(
            session,
            team_member_id=member_id,
            date=date,
            bucket=DAY_BUCKET,
            capacity=member["dailyCapacity"],
            booking_id=booking_id,
        )
        if shift_id:
            _claim_bucket(
                session,
                team_member_id=member_id,
                date=date,
                bucket=shift_id,
                capacity=shift_capacity if shift_capacity is not None else member["dailyCapacity"],
                booking_id=booking_id,
            )


def _claim_bucket(
    session: Session,
    *,
    team_member_id: str,
    date: str,
    bucket: str,
    capacity: int,
    booking_id: str,
) -> None:
    if capacity <= 0:
        raise CapacityFullError()
    existing = session.scalars(
        select(CapacityHold).where(
            CapacityHold.team_member_id == team_member_id,
            CapacityHold.date == date,
            CapacityHold.bucket == bucket,
        )
    ).all()
    taken = {row.seat for row in existing}
    for seat in range(capacity):
        if seat in taken:
            continue
        try:
            with session.begin_nested():
                session.add(
                    CapacityHold(
                        id=new_id(),
                        team_member_id=team_member_id,
                        date=date,
                        bucket=bucket,
                        seat=seat,
                        booking_id=booking_id,
                    )
                )
                session.flush()
            return
        except Exception as error:
            if is_unique_constraint(error):
                continue
            raise
    raise CapacityFullError()


def remaining_of(staff: dict) -> int:
    return max(0, min(staff["remainingDay"], staff["remainingShift"]))


def used_seats(session: Session, *, team_member_id: str, date: str, bucket: str) -> int:
    rows = session.scalars(
        select(CapacityHold).where(
            CapacityHold.team_member_id == team_member_id,
            CapacityHold.date == date,
            CapacityHold.bucket == bucket,
        )
    ).all()
    return len(rows)


def staff_snapshots(session: Session, business, date: str, shift_id: str | None) -> list[dict]:
    shift = next((row for row in business.shifts if row.id == shift_id and row.active), None) if shift_id else None
    members = [member for member in business.members if member.status == "ACTIVE"]
    rows = []
    for member in members:
        used_day = used_seats(session, team_member_id=member.id, date=date, bucket=DAY_BUCKET)
        used_shift = used_seats(session, team_member_id=member.id, date=date, bucket=shift.id) if shift else 0
        shift_cap = shift.capacity if shift and shift.capacity is not None else member.daily_capacity
        rows.append(
            {
                "id": member.id,
                "name": member.name,
                "dailyCapacity": member.daily_capacity,
                "shiftCapacity": shift.capacity if shift else None,
                "serviceIds": [item.service_id for item in member.services],
                "remainingDay": max(0, member.daily_capacity - used_day),
                "remainingShift": max(0, shift_cap - used_shift),
            }
        )
    return rows


def member_can_perform(member, service_id: str, kind: str | None = None) -> bool:
    from bridey_api.roles import member_matches_service_kind, parse_roles

    if kind:
        return member_matches_service_kind(parse_roles(member.roles), kind)
    return any(row.service_id == service_id for row in member.services)


def assign_staff(
    *,
    service_ids: list[str],
    staff: list[dict],
    preferred_member_id: str | None = None,
    preferred_by_service: dict | None = None,
    business,
) -> list[dict]:
    from bridey_api.errors import PreferredUnavailableError

    remaining = {row["id"]: remaining_of(row) for row in staff}
    assigned_this_booking: set[str] = set()
    result: list[dict] = []

    def kind_of(service_id: str) -> str | None:
        service = next((row for row in business.services if row.id == service_id), None)
        return service.kind if service else None

    def capable(service_id: str):
        return [
            row
            for row in staff
            if (member := next((item for item in business.members if item.id == row["id"]), None))
            and member_can_perform(member, service_id, kind_of(service_id))
        ]

    def take_member(member_id: str, service_id: str) -> None:
        if member_id not in assigned_this_booking:
            left = remaining.get(member_id, 0)
            if left < 1:
                raise CapacityFullError()
            remaining[member_id] = left - 1
            assigned_this_booking.add(member_id)
        result.append({"serviceId": service_id, "teamMemberId": member_id})

    preferred_map: dict[str, str] = {}
    if preferred_by_service:
        for service_id in service_ids:
            member_id = (preferred_by_service.get(service_id) or "").strip()
            if member_id:
                preferred_map[service_id] = member_id
    preferred = next((row for row in staff if row["id"] == preferred_member_id), None) if preferred_member_id else None
    if preferred_member_id and not preferred:
        raise PreferredUnavailableError()
    if preferred:
        for service_id in service_ids:
            member = next((item for item in business.members if item.id == preferred["id"]), None)
            if not preferred_map.get(service_id) and member and member_can_perform(member, service_id, kind_of(service_id)):
                preferred_map[service_id] = preferred["id"]

    leftover: list[str] = []
    for service_id in service_ids:
        preferred_id = preferred_map.get(service_id)
        if not preferred_id:
            leftover.append(service_id)
            continue
        snapshot = next((row for row in staff if row["id"] == preferred_id), None)
        biz_member = next((item for item in business.members if item.id == preferred_id), None)
        if not snapshot or not biz_member or not member_can_perform(biz_member, service_id, kind_of(service_id)):
            raise PreferredUnavailableError()
        if remaining_of(snapshot) < 1 and snapshot["id"] not in assigned_this_booking:
            raise PreferredUnavailableError()
        try:
            take_member(snapshot["id"], service_id)
        except CapacityFullError as error:
            raise PreferredUnavailableError() from error

    for service_id in leftover:
        options = capable(service_id)
        if not options:
            raise CapacityFullError("NO_STAFF")
        ranked = sorted(
            options,
            key=lambda row: (
                -(999 if row["id"] in assigned_this_booking else remaining.get(row["id"], 0)),
                row["name"],
            ),
        )
        pick = next((row for row in ranked if row["id"] in assigned_this_booking or remaining.get(row["id"], 0) > 0), None)
        if not pick:
            raise CapacityFullError()
        take_member(pick["id"], service_id)
    return result


def remaining_bookings_for_services(business, staff: list[dict], service_ids: list[str]) -> int:
    count = 0
    clone = [dict(row) for row in staff]
    for _ in range(40):
        try:
            assigned = assign_staff(service_ids=service_ids, staff=clone, business=business)
            used = {row["teamMemberId"] for row in assigned}
            for member_id in used:
                row = next((item for item in clone if item["id"] == member_id), None)
                if not row:
                    continue
                row["remainingDay"] = max(0, row["remainingDay"] - 1)
                row["remainingShift"] = max(0, row["remainingShift"] - 1)
            count += 1
        except (CapacityFullError, PreferredUnavailableError):
            break
    return count
